src/agent.py

```diff
@@ -127,26 +191,51 @@ def run(self):
 
             # Task Complete
             if tool_name == "task_complete":
-                action_result = args["result"]
+                action_result = args["result"]  # todo: check if result is correct
+                # todo: check if list of TASK need be updated
                 self._task_complete(action_result)
                 # save agent data
                 with self.ui.loading("Save agent data..."):
                     self.save_agent()
 
             # Action with tools
             else:
-                # Ask for permission to run the action
-                user_permission = self.ui.get_binary_user_input(
-                    "Do you want to continue?")
-                if not user_permission:
-                    action_result = "User Denied to run Action"
-                    self.ui.notify(title="USER INPUT", message=action_result)
-                else:
-                    try:
-                        action_result = self._act(tool_name, args)
-                    except Exception as e:
-                        raise e
-                    self.ui.notify(title="ACTION RESULT", message=action_result)
+                tool = None
+                try:
+                    tool = self.prodedural_memory.remember_tool_by_name(tool_name)
+
+                    if not tool:
+                        raise Exception(
+                            f"Tool {tool_name} is not found in the procedural memory.")
+
+                    def act(tool_name, args):
+                        try:
+                            action_result = self._act(tool_name, args)
+                        except Exception as e:
+                            raise e
+                        self.ui.notify(title="ACTION RESULT", message=action_result)
+                        return action_result
+
+                    self.ui.notify(
+                        title="LOG", message=f"Tool: {tool_name} user_permission_required:{tool.user_permission_required}")
+                    if tool.user_permission_required:
+                        # Ask for permission to run the action
+                        user_permission = self.ui.get_binary_user_input(
+                            "Do you want to continue?")
+                        if not user_permission:
+                            action_result = "User Denied to run Action"
+                            self.ui.notify(title="USER INPUT", message=action_result)
+                        else:
+                            action_result = act(tool_name, args)
+                    else:
+                        action_result = act(tool_name, args)
+                except ToolNotFoundException as e:
+                    self.ui.notify(
+                        title="LOG", message=f"Tool {tool_name} not exist. Pick one tool [TOOL] from the list.")
```

Comment:
For both line 234 and 235, adjust prompt to: f"Tool {tool_name} does not exist. Pick one tool [TOOL] from the list."